proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,205 @@
+Fine-Grained Unused Warnings
+============================
+
+.. author:: Jakob Brünker
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/434>`_.
+.. contents::
+
+This proposal aims to make warnings about unused names more precise.
+
+Motivation
+----------
+
+::
+
+  a :: Int
+  a = 4
+    where bar = foo
+          foo = 3
+
+Compiling the above code with ``-Wall`` results in the
+these warnings:
+
+::
+
+      Defined but not used: `bar'
+      Defined but not used: `foo'
+
+Sometimes, I look at a warning like this and think "I could have sworn I used
+``foo`` somewhere" - but of course, what the warning is trying to say isn't
+actually that ``foo`` is unused, but rather that all of ``foo``'s usage sites
+are in unused bindings.
+
+Furthermore, while it's useful to see such a warning for CI, to immediately
+know all of the bindings that have to be removed to eliminate the warnings,
+during development, these additional warnings take up a lot of valuable
+screen real estate in ghcid.
+
+Proposed Change Specification
+-----------------------------
+The overarching idea is to separate warnings about names that are genuinely
+unused, and warnings that are only "transitively unused" (or "indirectly unused"), i.e. only used by
+other (transitively) unused bindings.
+
+The warnings about transitively unused names will be reported if a new flag, `-freport-indirectly-unsed-bindings` is enabled (by default, it is) *and* the appropriate existing warning flags about unused bindings is enabled, i.e. one of
+
+* ``-Wunused-top-binds``
+* ``-Wunused-local-binds``
+* ``-Wunused-pattern-binds``
+* ``-Wunused-binds``
+* ``-Wunused-foralls``
+* ``-Wunused-matches``
+* ``-Wunused-imports``
+* ``-Wunused-type-patterns``
+
+In each of these cases, if a name is not used by anything, the same warning will be produced that has been produced prior to this proposal.
+
+For example, in the following code
+
+::
+
+  module Foo ()
+
+  import Data.List as L
+
+  foo = L.intercalate
+
+``intercalate`` is transitively unused, since it is only used by ``foo``, which is unused. The warning that's produced will be
+
+::
+
+  Foo.hs:5:1: warning: [-Wunused-top-binds]
+      Defined but not used: ‘foo’
+  
+  Foo.hs:3:1: warning: [-Wunused-imports, -freport-indirectly-used-bindings]:
+      The import of `Data.List' is only used by the following unused binding: ‘foo’
+        except perhaps to import instances from `Data.List'
+      To import instances alone, use: import Data.List()
+
+The warning about transitively unused bindings only appears if both the warning flag
+for the unused binding (in this case ``-Wunused-top-binds``) and the the one for
+the transitively unused binding itself (in this case ``-Wunused-imports``) are
+enabled.
```

Comment:
Good point, in this particular case it doesn't look like that would be a change from the current behavior, but in general it could happen that the number of warnings a code base produces changes, in cases where two warning flags interact. I've changed the text accordingly.